### DISCUSSION TITLE
Image Function Mesh Refinement Error when using with NS Module

### DISCUSSION THREAD

**Discussion by m1russell** [idaholab/moose#27571] [Q&A Modules: Navier-Stokes] [answered]
I am looking for some advice on some strange errors I am getting attempting  to use Moose's NS module to predict flow through voids from a 2 phase microCT image.  Many thanks!

I first use ImageFunction with a GradientJumpIndicator to create an adaptively refined mesh which I separate into two blocks using the approach suggested in this discussion: https://github.com/idaholab/moose/discussions/26733.

I then read in the refined mesh, remove the block corresponding to the solid-phase, and perform a NS analysis with unit pressure gradient (1 on the left face and zero on the right). The results of the analysis are shown below. It looks like the pressure is only being carried along certain paths were the elements are all equivalent sizes. 

![image](https://github.com/idaholab/moose/assets/142835798/7bc1a03d-39f5-4842-a744-a082c9af1794)

and here are results with no adaptive mesh refinement, just uniform refinement, 

![image](https://github.com/idaholab/moose/assets/142835798/210a41b2-1285-48bb-9e7e-36af50406633).

In addition, I had to play around with solver settings. Using the standard bjacobi preconditioner resulted in the following error which I resolved by using an asm precon. with an lu sub_pc following this page (https://mooseframework.inl.gov/modules/porous_flow/solvers.html) 
```

  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to SUBPC_ERROR
```

**Step 1: Adaptive Refinement Script**

````
[Mesh]
  [gen]
        type = GeneratedMeshGenerator
        dim = 2
        nx = 25 # = number of image pixels in x
        ny = 25 # = number of image pixels in y
    []
  [block0]
    type = SubdomainBoundingBoxGenerator
    input = 'gen'
    block_id = 0
    bottom_left = '0 0 0'
    top_right = '1 0.5 0'
  []
  [block1]
    type = SubdomainBoundingBoxGenerator
    input = 'block0'
    block_id = 1
    bottom_left = '0 0.5 0'
    top_right = '1 1 0'
  []
[]

[Functions]
    [image_func]
      type = ImageFunction
      file_base = test_images/Random_Sphers_25x_
      file_suffix = png
      file_range = '1'
      threshold = 1
      upper_value = 1
      lower_value = 0
    []
[]
  
[AuxVariables]
  [auxBlockID]
  []
[]

[AuxKernels]
  [aux_kern]
    type = FunctionAux
    variable = auxBlockID
    function = image_func
    execute_on = 'INITIAL'
  []
[]

[UserObjects]
  [change_blockID]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = auxBlockID
    criterion_type = ABOVE
    threshold = 0
    subdomain_id = 1
    complement_subdomain_id = 0
    execute_on = 'INITIAL'
  []
[]


[Variables]
    [u]
    []
[]

#Try use the auxillary variable in the adaptivity criterion instead of assigning it to the displacement variable
[ICs]
    [u_ic]
      type = FunctionIC
      function = image_func
      variable = u
    []
[]
  
[Problem]
    type = FEProblem
    solve = false  # that was the other way no to solve the problem, but only generate the mesh
[]
  
[Executioner]
    type = Transient
    num_steps = 1
    dt = 0.1
[]
  
[Adaptivity]
    max_h_level = 2
    initial_steps = 2
    initial_marker = marker
    [Indicators]
      [indicator]
        type = GradientJumpIndicator
        variable = u
      []
    []
    [Markers]
      [marker]
        type = ErrorFractionMarker
        indicator = indicator
        refine = 0.9
      []
    []
[]
  
[Outputs]
    exodus = true
[]
````

**Step 2: NS Solution Script**

````
mu = 1.1
rho = 1

[Mesh]

    [adapt]
        type = FileMeshGenerator
        file =adapt_mesh_creation_out.e
    []

    [ed0]
        type = BlockDeletionGenerator
        input = adapt
        block = '1'
        new_boundary ='void_surface'
    []



[]

[Modules]
  [NavierStokesFV]
    compressibility = 'incompressible'
    porous_medium_treatment = true

    density = 'rho'
    dynamic_viscosity = 'mu'
    porosity = 'funcporosity'

    initial_velocity = '1e-6 1e-6 0'
    initial_pressure = 0.0

    inlet_boundaries = 'left'
    momentum_inlet_types = 'fixed-pressure'
    momentum_inlet_function = '1'
    outlet_boundaries = 'right'
    momentum_outlet_types = 'fixed-pressure'
    pressure_function = '0'

    wall_boundaries = 'top bottom void_surface '
    momentum_wall_types = 'noslip noslip noslip'
    

    friction_types = 'darcy '
    friction_coeffs = 'Darcy_coefficient'

    mass_advection_interpolation = 'average'
    momentum_advection_interpolation = 'average'
  []
[]

[FunctorMaterials]
  [const]
    type = GenericFunctorMaterial
    prop_names = 'rho mu'
    prop_values = '${rho} ${mu}'
    outputs = exodus
  []
  [friction]
    type = GenericVectorFunctorMaterial
    prop_names = 'Darcy_coefficient '
    prop_values = '0.1 0.1 0.1'
    outputs = exodus
  []

  [jump]
    type = PiecewiseByBlockFunctorMaterial
    prop_name = 'funcporosity'
    subdomain_to_prop_value = '0 0.1
                               1 0.1'
    outputs = exodus
  []

[]



[Executioner]
    type = Transient
    dt = 1.0
    solve_type = 'PJFNK'
    steady_state_detection = true
    steady_state_tolerance = 1E-8
    petsc_options_iname = ' -pc_type -pc_asm_overlap  -sub_pc_type -sub_pc_factor_shift_type -ksp_type  '
    petsc_options_value = 'asm 2 lu NONZERO gmres'
    nl_rel_tol = 1e-11
    nl_abs_tol = 1e-14
  []
  
  
  [Outputs]
    exodus = true
    print_linear_residuals = true 
    print_nonlinear_residuals = true
    #print_perf_log = true 
  
  []
````

Random_Spheres_25x_01.png: 
![Random_Sphers_25x_01](https://github.com/idaholab/moose/assets/142835798/063df8a2-6713-4528-bdf8-d2cb2fc2c229)

**Comment by GiudGiud** [comment]
> I then read in the refined mesh
>         file =adapt_mesh_creation_out.e

we cant read exodus files with adaptivity refinement patterns. You ll have to use the Checkpoint format (for both writing in the first simulation and reading in the second) so that we read it properly

if you used a MeshDiagnosticsGenerator on your final mesh you would see it was read with tons of non-conformalities.

**Reply by m1russell** [reply]
Thanks Guillaume! Is there an example of this beyond what's available here: https://mooseframework.inl.gov/source/outputs/Checkpoint.html

Also is there a way I can do this all in one script that I am just not understanding?

**Reply by GiudGiud** [reply, marked as answer]
It's just writing then reading. You can look at test/tests/checkpoint, maybe there's an example there?
To restart the variables from a checkpoint, use the parameters in the problem (restart_file_base notably)

But yes you are right. You could do this in one step. Just use a step of 'initial' adaptivity instead of regular adaptivity.
The parameters for initial adaptivity are specified in the same block
